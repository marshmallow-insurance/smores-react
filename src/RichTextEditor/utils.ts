import { $createParagraphNode, LexicalNode, RootNode } from 'lexical'

export const appendNodes = (root: RootNode, nodes: LexicalNode[]) => {
  nodes
    .filter((node) => node.__type !== 'linebreak')
    .map((node) => {
      if (node.__type === 'text') {
        const paragraphNode = $createParagraphNode()
        paragraphNode.append(node)
        return paragraphNode
      }
      return node
    })
    .forEach((node) => root.append(node))
}
