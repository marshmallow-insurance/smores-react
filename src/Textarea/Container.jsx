import React, {useState} from 'react';
import {text, boolean} from '@storybook/addon-knobs';

import {Textarea} from './Textarea';

export const Container = () => {
  const [value, setValue] = useState('');

  return (
    <form>
      <Textarea
        id="textarea_id"
        label={text('label', 'Textarea label')}
        value={value}
        resize="both"
        onChange={setValue}
        disabled={boolean('disabled', false)}
        error={value.length > 255}
        errorMsg={`Your text is ${value.length} characters long. The maximum is 255 characters. Please make it shorter`}
        placeholder="Here is some placeholder text."
      />
    </form>
  );
};
