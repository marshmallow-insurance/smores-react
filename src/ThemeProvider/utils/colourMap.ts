import * as designTokens from '@mrshmllw/smores-foundations/build/web/variables.json'
import { Color } from 'theme'

type Flatten<T, Prefix extends string = ''> = {
  [K in keyof T & string]: T[K] extends Record<string, any>
    ? Flatten<T[K], `${Prefix}${K}.`>
    : `${Prefix}${K}`
}[keyof T & string]

export type NewColor = Flatten<
  Pick<typeof designTokens, 'color' | 'extended1' | 'thirdParty'>
>

export type ColorTypes = Color | NewColor

// TODO: Test how well this works
// we currently access numerical colours like `theme.color.neutral[100]` or in the case of 000 colours, `theme.color.neutral['000']`
export const getThemeColor = (path: NewColor): string => {
  return path
    .split('.')
    .reduce((acc, key) => acc?.[key], designTokens as any) as string
}

// a function that returns a flattened dot notation string path to access the color value
export const formatDesignTokenColor = (): string => {
  const { color, extended1, thirdParty } = designTokens
  const colors = { color, extended1, thirdParty }

  if (colors === null) {
    return ''
  }

  const result: string[] = []
  const destructureNestedObject = (obj: any, currentPath: string) => {
    for (const key in obj) {
      if (Object.prototype.hasOwnProperty.call(obj, key)) {
        const value = obj[key]
        const newPath = currentPath ? `${currentPath}.${key}` : key
        if (typeof value === 'object' && value !== null) {
          destructureNestedObject(value, newPath)
        } else {
          result.push(newPath)
        }
      }
    }
  }
  destructureNestedObject(colors, '')
  return result.join(', ')
}

// old colour name â†’ new colour path // base token value
export const legacyColorMap: Record<Color, NewColor> = {
  fairyFloss: 'color.surface.brand.100', // palette.primary.040
  bubblegum: 'color.surface.brand.200', // palette.primary.060
  marshmallowPink: 'color.surface.brand.300', // palette.primary.100
  lollipop: 'color.surface.brand.400', // palatte.primary.120

  // Core Secondary
  chia: 'color.feedback.inactive.100', // palette.secondary.040
  sesame: 'color.text.subtle', // palette.secondary.060
  liquorice: 'color.text.base', // palette.secondary.100
  boba: 'color.text.contrast', // palette.secondary.120

  // Core Tertiary
  cream: `color.surface.base.000`, // palette.neutral.000
  coconut: 'color.surface.base.100', // palette.neutral.010
  mascarpone: 'color.surface.base.200', // palette.neutral.020
  custard: 'color.surface.base.300', // palette.neutral.040

  // Brand Secondary
  feijoa: 'color.illustration.accent1.100', // palette.brand1.060
  spearmint: 'color.illustration.accent1.200', // palette.brand1.100
  macaroon: 'color.illustration.accent2.100', // palette.brand2.060
  blueberry: 'color.illustration.accent2.200', // palette.brand2.100
  pistachio: 'color.illustration.accent3.200', // palette.brand3.100
  matcha: 'color.illustration.accent3.100', // palette.brand3.060
  caramel: 'color.illustration.accent4.200', // palette.brand4.100
  peanut: 'color.illustration.accent4.100', // palette.brand4.060
  marzipan: 'color.illustration.neutral.400', // palette.neutral.100
  oatmeal: 'color.illustration.neutral.300', // palette.neutral.060
  satsuma: 'extended1.20', // extended1.020

  // Traffic light
  watermelon: 'color.feedback.negative.100', // palette.negative.020
  strawberry: 'color.feedback.negative.200', // palette.negative.100
  apple: 'color.feedback.positive.200', // palette.positive.100
  mint: 'color.feedback.positive.100', // palette.positive.020
  lemon: 'color.feedback.notice.200', // palette.notice.100
  sherbert: 'color.feedback.notice.100', // palette.notice.020
  tangerine: 'extended1.100', // extended1.100

  // Third-party brand colours
  compareTheMarket: 'thirdParty.compareTheMarket',
  confused: 'thirdParty.confusedCom',
  onfido: 'thirdParty.onfido',
  x: 'thirdParty.twitter', // x rebrand not yet reflected in foundations
  premfina: 'thirdParty.premfina',
  checkout: 'thirdParty.checkout',
  meta: 'thirdParty.facebook', // Meta rebrand not yet reflected in foundations
  stripe: 'thirdParty.stripe',
  intercom: 'thirdParty.intercom',
  ravelin: 'thirdParty.ravelin',
  rac: 'thirdParty.rac',
  hometree: 'thirdParty.hometree',
} as const

/* usage example */
// prop change:
// change color prop type to be `color?: Color | NewColor`
// component usage:
// <Component color={'marshmallowPink'}/>
// OR
// <Component color={'interactive.primary.base'}/>
// inside component:
// const resolvedColor = color in legacyColorMap ? getThemeColor(legacyColorMap[color]) : getThemeColor(color)

// the above should resolve to #FF88C8
