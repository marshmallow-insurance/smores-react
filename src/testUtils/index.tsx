import React, { ReactElement } from 'react'
import { render, RenderOptions } from '@testing-library/react'
import { SmoresThemeProvider } from '../ThemeProvider/ThemeProvider'

const Providers = ({ children }: { children: React.ReactNode }) => {
  return <SmoresThemeProvider>{children}</SmoresThemeProvider>
}

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, 'wrapper'>,
) => render(ui, { wrapper: Providers, ...options })

export * from '@testing-library/react'
export * from '@testing-library/jest-dom/vitest'
export * from 'jest-styled-components'

export { customRender as render }
