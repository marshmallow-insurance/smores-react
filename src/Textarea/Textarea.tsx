import React, { FormEvent, ForwardedRef, forwardRef } from 'react'
import styled from 'styled-components'
import { darken } from 'polished'

import { theme } from '../theme'
import { useUniqueId } from '../utils/id'
import { Field } from '../Field'

type BaseProps = {
  id?: string
  className?: string
  placeholder?: string
  label?: string
  name?: string
  value: string
  error?: boolean
  errorMsg?: string
  resize?: 'none' | 'both'
  disabled?: boolean
  maxLength?: number
  onBlur?: (e: FormEvent<HTMLTextAreaElement>) => void
  rows?: number
  required?: boolean
  renderAsTitle?: boolean
}

type TruncateProps =
  | {
      onChange: (e: string) => void
      onInputChange?: (e: FormEvent<HTMLTextAreaElement>) => void
    }
  | {
      onChange?: (e: string) => void
      onInputChange: (e: FormEvent<HTMLTextAreaElement>) => void
    }

export type TextareaProps = BaseProps & TruncateProps

export const Textarea = forwardRef(function Textarea(
  {
    id: idProp,
    name,
    className = '',
    label,
    value,
    onChange,
    onInputChange,
    resize = 'none',
    error = false,
    errorMsg,
    placeholder,
    disabled = false,
    maxLength,
    onBlur,
    rows = 4,
    required = false,
    renderAsTitle = false,
  }: TextareaProps,
  ref: ForwardedRef<HTMLTextAreaElement>,
) {
  const id = useUniqueId(idProp)
  return (
    <Field
      className={className}
      renderAsTitle={renderAsTitle}
      id={id}
      error={error}
      label={label}
      value={value}
      errorMsg={errorMsg}
      required={required}
      fullHeight
    >
      <StyledTextArea
        error={error}
        id={id}
        name={name}
        disabled={disabled}
        resize={resize}
        placeholder={placeholder}
        value={value}
        onChange={(e: FormEvent<HTMLTextAreaElement>) => {
          onChange?.(e.currentTarget.value)
          onInputChange?.(e)
        }}
        maxLength={maxLength}
        ref={ref}
        onBlur={onBlur}
        rows={rows}
      />
    </Field>
  )
})

interface TextArea {
  resize: 'none' | 'both'
  disabled: boolean
  error: boolean
  value: string
}

const StyledTextArea = styled.textarea<TextArea>`
  font-size: 16px;
  line-height: 20px;
  background: ${theme.colors.white};
  border: 2px solid ${theme.colors.outline};
  box-sizing: border-box;
  border-radius: 8px;
  width: 100%;
  padding: 16px;
  color: ${theme.colors.secondary};
  resize: ${({ resize }) => resize};
  cursor: ${({ disabled }) => (disabled ? 'not-allowed' : 'text')};
  opacity: ${({ disabled }) => (disabled ? '0.5' : '1')};
  border-color: ${({ error }) =>
    theme.colors[`${error ? 'error' : 'outline'}`]};
  outline: none;

  &:hover,
  &:focus,
  &:focus-visible {
    border-color: ${({ error }) =>
      error ? theme.colors.error : darken(0.1, theme.colors.outline)};
  }

  ${({ value }) =>
    value &&
    value != '' &&
    `
      border-color: ${theme.colors.outline};
    `}
`
