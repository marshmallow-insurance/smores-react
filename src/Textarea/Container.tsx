import React, { useState } from 'react'

import { Textarea } from './Textarea'

export const Container = () => {
  const [value, setValue] = useState('')

  return (
    <form>
      <Textarea
        id="textarea_id"
        label="Textarea label"
        value={value}
        resize="both"
        onChange={setValue}
        assistiveText="This is some assistive text"
        disabled={false}
        error={value.length > 255}
        errorMsg={`Your text is ${value.length} characters long. The maximum is 255 characters. Please make it shorter`}
        placeholder="Here is some placeholder text."
      />
    </form>
  )
}
