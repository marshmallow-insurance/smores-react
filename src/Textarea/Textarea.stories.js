import React from 'react'
import { Container } from './Container'
import { Textarea } from './Textarea'

export default {
  title: 'Textarea',
  component: Textarea,
}

const Template = (args) => <Textarea {...args} />

export const Default = Template.bind({})

Default.args = {
  id: 'textarea_id',
  label: 'Textarea label',
  value: '',
  onChange: () => {},
  disabled: false,
  placeholder: 'Here is some placeholder text.',
}

export const Disabled = Template.bind({})

Disabled.args = {
  id: 'textarea_id',
  label: 'Textarea label',
  value: '',
  onChange: () => {},
  resize: 'none',
  disabled: true,
  placeholder: 'Here is some placeholder text.',
}

export const Error = Template.bind({})

Error.args = {
  id: 'textarea_id',
  label: 'Textarea label',
  value: '',
  onChange: () => {},
  disabled: false,
  placeholder: 'Here is some placeholder text.',
  resize: 'none',
  error: true,
  errorMsg: 'Something really quite terrible has gone wrong here!',
}

const WorkingExampleTemplate = (args) => <Container {...args} />

export const WorkingExample = WorkingExampleTemplate.bind({})
