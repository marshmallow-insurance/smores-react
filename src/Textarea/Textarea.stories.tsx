import React from 'react'
import { noop } from '../utils/noop'
import { Container } from './Container'
import { Textarea, TextareaProps } from './Textarea'

export default {
  title: 'Textarea',
  component: Textarea,
}

const Template = (props: TextareaProps) => <Textarea {...props} />

export const Default = Template.bind({})

Default.args = {
  id: 'textarea_id',
  label: 'Textarea label',
  value: '',
  onChange: noop,
  disabled: false,
  placeholder: 'Here is some placeholder text.',
}

export const Disabled = Template.bind({})

Disabled.args = {
  id: 'textarea_id',
  label: 'Textarea label',
  value: '',
  onChange: noop,
  resize: 'none',
  disabled: true,
  placeholder: 'Here is some placeholder text.',
}

export const Error = Template.bind({})

Error.args = {
  id: 'textarea_id',
  label: 'Textarea label',
  value: '',
  onChange: noop,
  disabled: false,
  placeholder: 'Here is some placeholder text.',
  resize: 'none',
  error: true,
  errorMsg: 'Something really quite terrible has gone wrong here!',
}

export const Required = Template.bind({})

Required.args = {
  id: 'textarea_id',
  label: 'Textarea label',
  value: '',
  onChange: noop,
  disabled: false,
  placeholder: 'Here is some placeholder text.',
  required: true,
}

export const WithTitleAsLabel = Template.bind({})

WithTitleAsLabel.args = {
  id: 'textarea_id',
  label: 'Textarea label',
  value: '',
  onChange: noop,
  disabled: false,
  placeholder: 'Here is some placeholder text.',
  required: false,
  renderAsTitle: true,
}

const WorkingExampleTemplate = () => <Container />

export const WorkingExample = WorkingExampleTemplate.bind({})
