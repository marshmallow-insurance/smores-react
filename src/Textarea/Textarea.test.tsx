import React from 'react'
import { render, screen } from '../testUtils'
import { Textarea } from './Textarea'
import { noop } from '../utils/noop'

describe('Textarea', () => {
  it('renders correctly with required props', () => {
    const { baseElement } = render(
      <Textarea value="Textarea label" onChange={noop} />,
    )

    expect(baseElement).toMatchSnapshot()

    expect(screen.getByText('Textarea label')).toBeTruthy()
  })
})
