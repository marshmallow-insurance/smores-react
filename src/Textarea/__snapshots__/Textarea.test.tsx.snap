// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Textarea > renders correctly with required props 1`] = `
<body>
  <div>
    <div
      class="sc-bRKDuR bHjpxP sc-jJLAfE izsDDD"
    >
      <div
        class="sc-bRKDuR bHjpxP"
      >
        <textarea
          class="sc-hwkwBN ipGeao"
          id="MM_SMORES_0"
          rows="4"
        >
          Textarea label
        </textarea>
      </div>
    </div>
  </div>
</body>
`;
